File: leetcode-java/src/main/java/chapter02/ReorderList.java
package chapter02;

import utils.ListNode;

/**
 * <h2>重排链表</h2>
 */
public class ReorderList {

    /**
     * <p>思路: 找中点 + 逆序 + 衔接</p>
     * <p>注: 几乎考察了链表常用的几个技巧</p>
     */
    private static ListNode reorderList(ListNode head){
        ListNode next = null;
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        fast = head;
        slow = ReverseList.reverseListLoop(slow);
        while (slow.next != null) {
            next = fast.next;
            fast.next = slow;
            fast = next;
            next = slow.next;
            slow.next = fast;
            slow = next;
        }
        return head;
    }
}

File: leetcode-go/chapter02/reorder_list.go
package main

import "leetcode-go/node"

type ListReorder struct {}

// 重排链表
func (reorder *ListReorder) ReorderList(head *node.ListNode) *node.ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	fast = head
	reverser := &ListReverser{}
	slow = reverser.ReverseListLoop(slow)
	for slow.Next != nil {
		fastNext := fast.Next
		slowNext := slow.Next
		fast.Next = slow
		slow.Next = fastNext
		fast = fastNext
		slow = slowNext
	}
	return head
}
